src/adapter/src/catalog/transact.rs

```diff
@@ -2262,6 +2268,137 @@ impl Catalog {
         Ok(())
     }
 
+    /// Verify retain history requirements of `id` (an index on a view must have a retain history >=
+    /// the retain history of any view index that uses it). If `id` is not an Index on a View,
+    /// returns Ok.
+    ///
+    /// The intention of this function is to prevent unexpected downstream memory growth. When the
+    /// retain history is increased on an index that is on a view, any indexes that are used by that
+    /// index have their compaction window implicitly increased. This could greatly and unexpectedly
+    /// increase memory use of those downstream indexes. To prevent this surprise to users, require
+    /// that users first increase all downstream indexes so that they are aware of and opting in to
+    /// the increased memory usage. If a persist object (table, materialized view, source) is used
+    /// by the index, we do not need to check further because the increased retain history of the
+    /// index will only result in a longer persist compaction window resulting in more disk use, but
+    /// not more memory use. Because disk is cheap and can't run out, it is ok for this increase to
+    /// not be explicit by the user.
+    fn check_index_retain_histories(
+        catalog: &CatalogState,
+        id: &GlobalId,
+    ) -> Result<(), AdapterError> {
+        let entry = catalog.get_entry(id);
+        let Some(index) = entry.index() else {
+            return Ok(());
+        };
+        let on = catalog.get_entry(&index.on);
+        if !on.is_view() {
+            // Non-views don't have this cascade requirement.
+            return Ok(());
+        }
+        let index_rh = index.custom_logical_compaction_window.unwrap_or_default();
+        for uses_index_id in Self::get_view_index_uses(catalog, id) {
```

Comment:
Is it possible for this set of indexes to change in-between version upgrades which could potentially cause this check to fail after a version upgrade when it was succeeding previously?

Reply:
No, the indexes here are user indexes in the catalog. There's no logic around dataflow index selection involved, so this should be the same across version upgrades.